sqlmesh/core/snapshot/definition.py

```diff
@@ -423,6 +425,13 @@ class SnapshotTableInfo(PydanticModel, SnapshotInfoMixin, frozen=True):
     # This can be removed from this model once Pydantic 1 support is dropped (must remain in `Snapshot` though)
     base_table_name_override: t.Optional[str] = None
 
+    custom_materialization: t.Optional[str] = None
```

Comment:
why is this needed on the table info?

Reply:
Because in snapshot evaluator we don't always have the full snapshot available. Particularly during demotion and cleanup.

Reply:
in demotion, why do you need the custom materialization? do you know what the materialization is for a non custom materialization is for demotion/ cleanup?

Reply:
> do you know what the materialization is for a non custom materialization is for demotion/ cleanup?

we do based on the kind name

> in demotion, why do you need the custom materialization?

eg. I could decide to use branches in LakeFS or Iceberg, etc. instead of views. Even if not for demotion, we still need this for cleanup.

Reply:
so then don't you only need kind name, why do you also need properties?

should we just have kind_name be the custom kind name instead of custom?

Reply:
I'd rather not mix kind name and the materialization name. They are different concepts. We need properties to correctly initialize the custom materialization class instance.

Reply:
why do custom materializations need properties for initialization but not regular materializations?

Reply:
Because in regular materializations the logic is hardcoded. Properties provide additional flexibility for developers to reuse the same implementation with slightly different settings in different models.

Reply:
ok

Reply:
Also builtin materialization use the model definition directly and it just happens that the cleanup and demotion behavior don't depend on a specific model.

Reply:
Actually, I think that custom materialization properties can be sourced from the model definition itself which is available in all cases except demotion and cleanup. This doesn't seem like a big deal but makes the implementation cleaner. So I'll probably go ahead and change that.

Reply:
Done that